cloudnew/ec2_client.go

```diff
@@ -0,0 +1,193 @@
+	if creds == nil {
+		return errors.New("creds must not be nil")
+	}
+	if c.session == nil {
+		s, err := session.NewSession()
+		if err != nil {
+			return err
+		}
+		c.session = s
+	}
+	c.EC2 = ec2.New(c.session)
+	return nil
+}
+
+// RunInstances is a wrapper for ec2.RunInstances.
+func (c *AWSClientImpl) RunInstances(input *ec2.RunInstancesInput) (*ec2.Reservation, error) {
+	return c.EC2.RunInstances(input)
+}
+
+// DescribeInstances is a wrapper for ec2.DescribeInstances
+func (c *AWSClientImpl) DescribeInstances(input *ec2.DescribeInstancesInput) (*ec2.DescribeInstancesOutput, error) {
+	return c.EC2.DescribeInstances(input)
+}
+
+// CreateTags is a wrapper for ec2.CreateTags.
+func (c *AWSClientImpl) CreateTags(input *ec2.CreateTagsInput) error {
+	_, err := c.EC2.CreateTags(input)
+	return err
+}
+
+// TerminateInstances is a wrapper for ec2.TerminateInstances.
+func (c *AWSClientImpl) TerminateInstances(input *ec2.TerminateInstancesInput) (*ec2.TerminateInstancesOutput, error) {
+	return c.EC2.TerminateInstances(input)
+}
+
+// RequestSpotInstances is a wrapper for ec2.RequestSpotInstances.
+func (c *AWSClientImpl) RequestSpotInstances(input *ec2.RequestSpotInstancesInput) (*ec2.RequestSpotInstancesOutput, error) {
+	return c.EC2.RequestSpotInstances(input)
+}
+
+// DescribeSpotInstanceRequests is a wrapper for ec2.DescribeSpotInstanceRequests.
+func (c *AWSClientImpl) DescribeSpotInstanceRequests(input *ec2.DescribeSpotInstanceRequestsInput) (*ec2.DescribeSpotInstanceRequestsOutput, error) {
+	return c.EC2.DescribeSpotInstanceRequests(input)
+}
+
+// CancelSpotInstanceRequests is a wrapper for ec2.CancelSpotInstanceRequests.
+func (c *AWSClientImpl) CancelSpotInstanceRequests(input *ec2.CancelSpotInstanceRequestsInput) error {
+	_, err := c.EC2.CancelSpotInstanceRequests(input)
+	return err
+}
+
+// AWSClientMock mocks ec2.EC2.
+type AWSClientMock struct {
+	*credentials.Credentials
+	*ec2.RunInstancesInput
+	*ec2.DescribeInstancesInput
+	*ec2.CreateTagsInput
+	*ec2.TerminateInstancesInput
+	*ec2.RequestSpotInstancesInput
+	*ec2.DescribeSpotInstanceRequestsInput
+	*ec2.CancelSpotInstanceRequestsInput
+}
+
+// Create a new mock client.
+func (c *AWSClientMock) Create(creds *credentials.Credentials) error {
+	c.Credentials = creds
+	return nil
+}
+
+// RunInstances is a mock for ec2.RunInstances.
+func (c *AWSClientMock) RunInstances(input *ec2.RunInstancesInput) (*ec2.Reservation, error) {
+	c.RunInstancesInput = input
+	return &ec2.Reservation{
+		Instances: []*ec2.Instance{
+			&ec2.Instance{
+				InstanceId: makeStringPtr("instance_id"),
+				KeyName:    makeStringPtr("keyName"),
+			},
+		},
+	}, nil
+}
+
+// DescribeInstances is a mock for ec2.DescribeInstances
+func (c *AWSClientMock) DescribeInstances(input *ec2.DescribeInstancesInput) (*ec2.DescribeInstancesOutput, error) {
+	c.DescribeInstancesInput = input
+	return &ec2.DescribeInstancesOutput{
+		Reservations: []*ec2.Reservation{
+			&ec2.Reservation{
```

Comment:
no need to change it, but I think its valid to drop type names for slices, when its obvious)